### PR TITLE
fix(config_flow): support google validation

### LINKED ISSUE #20: Invalid authentication
there seems to be an issue with the authentication, I tried using my google generative ai key which works perfectly with the in-built google generative AI integration however I get `Invalid authentication` error while trying to setup the integration

### PR DESCRIPTION
**Pull request by grleblanc**
Fixes #20 

I tried looking around for an "easy" way to test our Google API Keys, however, after some serious digging this is what I came up with.

- Instructions can be found [here](https://aistudio.google.com/app/apikey)
- Will need to update the link in the documentation on where to get the API key

Example test here.

```
import requests

API_KEY = "foo/bar" # from https://aistudio.google.com/app/apikey

def test_validate_google_false() -> bool:
    """Validate Google configuration."""
    headers = {
        "Authorization": f"Bearer {API_KEY}",
        "Content-Type": "application/json",
    }
    try:
        print("Validating Google API key")
        # Placeholder URL; replace with the actual Google API endpoint
        response = requests.get("https://api.google.com/v1/models", headers=headers)
        print(response.status_code)
        is_valid = response.status_code == 200
        print("Google validation result: %s", is_valid)
        return is_valid
    except Exception as err:
        print("Google validation error: %s", err)
        return False

def test_validate_google_true() -> bool:
    """Validate Google configuration."""
    headers = {
        "Content-Type": "application/json",
    } 
    try:
        print("Validating Google API key")
        response = requests.post(
            f"https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key={API_KEY}",
            json={"contents": [{"parts": [{"text": "Explain how AI works"}]}]},
            headers=headers,
        )
        is_valid = response.status_code == 200
        print("Google validation result: %s", is_valid)
        return is_valid
    except Exception as err:
        print("Google validation error: %s", err)
        return False

if __name__ == "__main__":
    test_validate_google_false()
    test_validate_google_true()

>>

Validating Google API key
404
Google validation result: %s False

Validating Google API key
200
Google validation result: %s True
```

### CODE REVIEW

**Comment by ITSpecialist111** [comment]
Thank you for your suggestions; they seem to fix Google. It appears there are similar issues with Anthropic.

# Documentation of Changes Made to `config_flow.py`

---

## **Overview**

We have updated the `config_flow.py` file of the AI Automation Suggester integration to enhance error handling during the setup process. The key change is the implementation of capturing raw error messages from API responses and displaying them directly to the user during configuration. This approach provides users with clear and actionable information when issues arise, such as invalid authentication, insufficient credits, or other API-related errors.

---

## **Detailed Changes**

### **1. Modified Provider Validation Methods**

- **Return Type Update:**
  - **Before:** The validation methods returned a boolean (`True` for success, `False` for failure).
  - **After:** The methods now return an `Optional[str]`. They return `None` on success or an error message string on failure.

- **Error Handling Enhancement:**
  - Captured raw error messages from API responses.
  - Parsed the error responses to extract meaningful messages.
  - Logged detailed error information for debugging purposes.
  - Returned the error message to be displayed to the user during setup.

- **Example Change for Anthropic:**

  ```python
  async def validate_anthropic(self, api_key: str, model: str) -> Optional[str]:
      # ... existing headers and setup ...
      try:
          response = await self.session.post(
              "https://api.anthropic.com/v1/complete",
              headers=headers,
              json=payload
          )
          if response.status == 200:
              return None  # Success
          else:
              error_text = await response.text()
              _LOGGER.error(f"Anthropic validation error response: {error_text}")
              try:
                  error_json = await response.json()
                  error_message = error_json.get("error", {}).get("message", error_text)
              except Exception:
                  error_message = error_text
              return error_message
      except Exception as err:
          _LOGGER.error(f"Anthropic validation exception: {err}")
          return str(err)
  ```

- **Applied to All Providers:**
  - Similar changes were made to `validate_openai`, `validate_google`, `validate_groq`, `validate_localai`, `validate_ollama`, and `validate_custom_openai` methods.

### **2. Updated Configuration Steps**

- **Error Message Propagation:**
  - In each `async_step_*` method corresponding to the providers, updated error handling to pass the raw error message to the user interface.
  - If the validation method returns an error message, it is assigned to `errors["base"]` with a key of `"api_error"`, and the error message is passed using `description_placeholders`.

- **Example Change for Anthropic:**

  ```python
  async def async_step_anthropic(self, user_input: Optional[Dict[str, Any]] = None):
      errors = {}
      description_placeholders = {}
      if user_input is not None:
          # ... existing code to get model and validator ...
          error_message = await self.validator.validate_anthropic(api_key, model)
          if error_message is None:
              # Success
              self.data.update(user_input)
              return self.async_create_entry(
                  title="AI Automation Suggester (Anthropic)",
                  data=self.data
              )
          else:
              errors["base"] = "api_error"
              description_placeholders["error_message"] = error_message
      return self.async_show_form(
          step_id="anthropic",
          data_schema=vol.Schema({
              vol.Required(CONF_ANTHROPIC_API_KEY): str,
              vol.Optional(CONF_ANTHROPIC_MODEL, default=DEFAULT_MODELS["Anthropic"]): str,
          }),
          errors=errors,
          description_placeholders=description_placeholders
      )
  ```

- **Consistent Approach:**
  - Applied similar changes to all provider configuration steps (`async_step_openai`, `async_step_google`, etc.).

### **3. Updated `strings.json` for Error Messages**

- **Added New Error Message Key:**
  - Introduced an `"api_error"` key to display the error message passed from the validation method.

- **Example Addition:**

  ```json
  {
    "config": {
      "error": {
        "api_error": "API error: {error_message}",
        "already_configured": "This provider is already configured."
      }
    }
  }
  ```

- **Usage of Placeholders:**
  - `{error_message}` is a placeholder that gets replaced with the actual error message from the API.

### **4. Logging Enhancements**

- **Detailed Error Logging:**
  - In validation methods, logged the full error response and any exceptions that occur.
  - Helps in debugging issues by providing comprehensive logs.

- **Example Logging:**

  ```python
  _LOGGER.error(f"Anthropic validation error response: {error_text}")
  _LOGGER.error(f"Anthropic validation exception: {err}")
  ```

### **5. Consistency Across Providers**

- **Unified Error Handling:**
  - Ensured that all providers follow the same pattern for error capturing and displaying.
  - Simplifies maintenance and improves code readability.

---

## **Benefits of the Changes**

- **Improved User Experience:**
  - Users receive specific and actionable error messages during setup.
  - Reduces confusion by clearly indicating whether an issue is due to authentication, insufficient credits, or other API errors.

- **Enhanced Transparency:**
  - Users can see the exact error message returned by the API, helping them understand and resolve issues more efficiently.

- **Simplified Maintenance:**
  - A consistent error handling mechanism across all providers makes the codebase easier to manage and extend.

---

## **Usage Notes**

- **Sensitive Information:**
  - Be cautious with error messages to ensure they do not contain sensitive information such as API keys or personal data.
  - The error messages displayed are those returned by the APIs, so they should be safe, but always verify.

- **Testing:**
  - Thoroughly test the integration with various error scenarios (e.g., invalid API key, insufficient credits, network issues) to ensure error messages are displayed correctly.

- **Localization:**
  - Since raw error messages are in English (as returned by the APIs), consider providing translations or wrapping them in localized text if necessary.

---

## **Examples of Error Messages Displayed to Users**

- **Insufficient Credits:**
  - *"API error: Your credit balance is too low to access the Anthropic API. Please go to Plans & Billing to upgrade or purchase credits."*

- **Invalid API Key:**
  - *"API error: The API key provided is invalid. Please check your API key and try again."*

- **Network Error:**
  - *"API error: Cannot connect to the server. Please check your network connection and try again."*

---

## **Conclusion**

The changes made to the `config_flow.py` file significantly enhance the error handling capabilities of the AI Automation Suggester integration. By capturing and displaying raw error messages from API responses, users are better informed about issues that occur during setup and can take appropriate corrective actions. This leads to a more transparent and user-friendly experience.

**Comment by ITSpecialist111** [comment]
We would appreciate it if this code could be tested before committing. Here's the Config_flow file to replace the existing
[config_flow.py.txt](https://github.com/user-attachments/files/17604413/config_flow.py.txt)

**Comment by DUC750** [comment]
Anthropic doesn't work. "API error occurred"

**Comment by ITSpecialist111** [comment]
> Anthropic doesn't work. "API error occurred"

Can you check the logs as it should provide more information now.

File: custom_components/ai_automation_suggester/config_flow.py [hunk 1]
```diff
@@ -92,15 +92,21 @@ async def validate_anthropic(self, api_key: str) -> bool:
...
             _LOGGER.debug("Validating Google API key")
             # Placeholder URL; replace with the actual Google API endpoint
-            response = await self.session.get(
-                "https://api.google.com/v1/models",
-                headers=headers
+            response = await self.session.post(
+                "https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent",
+                headers=headers,
+                json=test_request,
+                params=param
```
Comment (Lash-L) [inline custom_components/ai_automation_suggester/config_flow.py:108]:
```suggestion
                params=params
```

Missing a s here!

**Comment by Lash-L** [comment]
> > Anthropic doesn't work. "API error occurred"
> 
> Can you check the logs as it should provide more information now.

At least for google, I am still getting errors - it seems error_text is blank.

**Comment by grleblanc** [comment]
With new config flow using Google API key created here: https://aistudio.google.com/app/apikey it works! Tested on 

```
Core
2024.10.4
Supervisor
2024.10.3
Operating System
13.2
Frontend
20241002.4
```

**Comment by DUC750** [comment]
> > > Anthropisch funktioniert nicht. "API-Fehler aufgetreten"
> > 
> > 
> > Können Sie die Protokolle überprüfen, da sie jetzt mehr Informationen liefern sollten.
> 
> Zumindest für Google bekomme ich immer noch Fehler - es scheint, dass error_text leer ist.

Sorry, it seems to be my fault. I thought the Free Plan from Anthropic would be enough.

`Zeile 562: 2024-11-02 11:49:30.332 ERROR (MainThread) [custom_components.ai_automation_suggester.config_flow] Anthropic validation error response: {"type":"error","error":{"type":"invalid_request_error","message":"Your credit balance is too low to access the Anthropic API. Please go to Plans & Billing to upgrade or purchase credits."}}`

**Comment by ITSpecialist111** [comment]
Yeah I know what you mean, the "free" APIs don't work as it requires the messaging APIs. I'll look into this though

**Comment by grleblanc** [comment]
I've also included a commit that implements the gemini-1.5-pro model, I have it all working end to end here now. The original code had code for a now deprecated PaLM endpoint (seen here https://ai.google.dev/api/palm).

We can switch to the gemini-1.5-flash model, that's fine too. Whatever you think is best.

**Comment by ITSpecialist111** [comment]
Spent the last few hours looking at the messages and testing the changes to the beta code above. I think we have a working version now after lots of testing. Thank you so much for your suggestions and testing!  I'll post V1.08 shortly.

**Comment by ITSpecialist111** [comment]
Updated to resolve issues within this pull request:  https://github.com/ITSpecialist111/ai_automation_suggester/releases/tag/V1.08

**Pull request closed without merging**